packages/yarnpkg-fslib/sources/path.ts

```diff
@@ -6,15 +6,40 @@ enum PathType {
   Native,
 }
 
-export type PortablePath = string & { _path_type: PathType.File | PathType.Portable };
-export type NativePath = string & { _path_type?: PathType.File | PathType.Native };
+enum PathRelativity {
+  Absolute,
+  Relative,
+}
+
+export type AbsolutePortablePath = string & {
+  _path_type: PathType.File | PathType.Portable;
+  _path_relativity: PathRelativity.Absolute;
+};
+export type AbsoluteNativePath = string & {
+  _path_type?: PathType.File | PathType.Native;
+  _path_relativity: PathRelativity.Absolute;
+};
+export type RelativePortablePath = string & {
+  _path_type: PathType.File | PathType.Portable;
+  _path_relativity: PathRelativity.Relative;
+};
+export type RelativeNativePath = string & {
+  _path_type?: PathType.File | PathType.Native;
+  _path_relativity: PathRelativity.Relative;
+};
+
+export type PortablePath = AbsolutePortablePath | RelativePortablePath;
+export type NativePath = AbsoluteNativePath | RelativeNativePath | (string & {_path_relativity?: void});
```

Comment:
`string & {_path_relativity?: void}` means all strings that don't have a relativity marker, which matches all strings apart from `{Absolute,Relative}{Native,Portable}Path`